File: 12-valid-parentheses/solution.cpp
// Valid Parentheses
// Time: O(n), Space: O(n)

#include <iostream>
#include <string>
#include <stack>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        unordered_map<char, char> mapping = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };
        
        for (char c : s) {
            if (mapping.count(c)) {
                char top = stk.empty() ? '#' : stk.top();
                if (!stk.empty()) stk.pop();
                if (mapping[c] != top) {
                    return false;
                }
            } else {
                stk.push(c);
            }
        }
        
        return stk.empty();
    }
};

int main() {
    Solution sol;
    cout << boolalpha;
    cout << sol.isValid("()[]{}") << endl;   // true
    cout << sol.isValid("([)]") << endl;     // false
    cout << sol.isValid("{[]}") << endl;     // true
    return 0;
}

File: 12-valid-parentheses/solution.cs
// Valid Parentheses
// Time: O(n), Space: O(n)

using System;
using System.Collections.Generic;

public class Solution {
    public bool IsValid(string s) {
        Stack<char> stack = new Stack<char>();
        Dictionary<char, char> mapping = new Dictionary<char, char> {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };
        
        foreach (char c in s) {
            if (mapping.ContainsKey(c)) {
                char top = stack.Count > 0 ? stack.Pop() : '#';
                if (mapping[c] != top) {
                    return false;
                }
            } else {
                stack.Push(c);
            }
        }
        
        return stack.Count == 0;
    }
    
    public static void Main(string[] args) {
        Solution sol = new Solution();
        Console.WriteLine(sol.IsValid("()[]{}"));   // True
        Console.WriteLine(sol.IsValid("([)]"));     // False
        Console.WriteLine(sol.IsValid("{[]}"));     // True
    }
}
